### PR TITLE
SPOI-8396 Create example for Kafka 0.9 API exactly-once output operator

### PR DESCRIPTION
**Pull request by oliverwnk**
@sandeshh please review and merge as appropriate

### CODE REVIEW

File: examples/kafka-exactly-once/src/main/java/com/example/KafkaTopicMessageInputOperator.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package com.example;
+
+import org.apache.apex.malhar.kafka.AbstractKafkaInputOperator;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.lib.util.KeyValPair;
+
+public class KafkaTopicMessageInputOperator extends AbstractKafkaInputOperator
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/main/java/com/example/KafkaTopicMessageInputOperator.java]: Why it is used?

File: examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      // run app asynchronously; terminate after results are checked
+      Configuration conf = getConfig();
+      LocalMode lma = LocalMode.newInstance();
+      lma.prepareDAG(new Application(), conf);
+      ValidationToFile validationToFile = (ValidationToFile)lma.getDAG().getOperatorMeta("validationToFile").getOperator();
+      LocalMode.Controller lc = lma.getController();
+      lc.runAsync();
+
+      // get messages from Kafka topic and compare with input
+      while(!validationToFile.validationDone)
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java]: Infinite loop if the validation fails.

File: examples/kafka-exactly-once/src/test/resources/log4j.properties [hunk 3]
```diff
@@ -0,0 +1,22 @@
+log4j.rootLogger=DEBUG,CONSOLE
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/resources/log4j.properties]: Debug generates lot of garbage, info preferrable.

File: examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        if (name.split("_")[0].equals("validation.txt")) {
+          return true;
+        }
+        return false;
+      }
+    };
+
+    File validationFile = folder.listFiles(filenameFilter)[0];
+
+    FileInputStream inputStream = new FileInputStream(validationFile);
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java]: use try-with-resources

File: examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    try {
+      validationOutput = IOUtils.toString(inputStream);
+      System.out.println("Validation output: " + validationOutput);
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java]: Don't use System.out.println, use logging everywhere.

File: examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      }
+      lc.shutdown();
+      Thread.sleep(1000);
+      checkOutput();
+    } catch (ConstraintViolationException e) {
+      Assert.fail("constraint violations: " + e.getConstraintViolations());
+    }
+  }
+
+  private void cleanup()
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java]: Who is calling cleanup? Better if it is done by Junit after/afterclass annotations.

File: examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  {
+    try {
+      FileUtils.deleteDirectory(new File(directory));
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/test/java/com/example/ApplicationTest.java]: Use deleteQuietly

File: examples/kafka-exactly-once/src/main/java/com/example/ValidationToFile.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      if (exactlyList == null) {
+        exactlyList = new ArrayList<>();
+      }
+      exactlyList.add(message);
+
+      processTuple(tuple);
+    }
+
+    @Override
+    public StreamCodec<byte[]> getStreamCodec()
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/main/java/com/example/ValidationToFile.java:79]: Why do you need to override this?

File: examples/kafka-exactly-once/src/main/java/com/example/ValidationToFile.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+  public final transient DefaultInputPort<byte[]> topicExactlyInput = new DefaultInputPort<byte[]>()
+  {
+    @Override
+    public void process(byte[] tuple)
+    {
+      String message = new String(tuple);
+      latestExactlyValue = message;
+      if (exactlyList == null) {
+        exactlyList = new ArrayList<>();
```
Comment (sandeshh) [inline examples/kafka-exactly-once/src/main/java/com/example/ValidationToFile.java]: initialization should be done in Setup

Comment (sandeshh) [reply examples/kafka-exactly-once/src/main/java/com/example/ValidationToFile.java]: Null check is done for every tuple, which is not required.

**Comment by tweise** [comment]
SPOI doesn't exist in Apex, please reference the appropriate JIRA.

**Pull request closed without merging**